src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/JsonPatchService.cs

```diff
@@ -0,0 +1,105 @@
+using Hl7.FhirPath;
+using Microsoft.AspNetCore.JsonPatch;
+using Microsoft.AspNetCore.JsonPatch.Exceptions;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Microsoft.Health.Fhir.Core.Extensions;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Models;
+
+namespace Microsoft.Health.Fhir.Core.Features.Resources.Patch
+{
+    public sealed class JsonPatchService
+    {
+        private readonly IModelInfoProvider _modelInfoProvider;
+
+        private readonly ISet<string> _immutableProperties = new HashSet<string>
+        {
+            "Resource.id",
+            "Resource.meta.lastUpdated",
+            "Resource.meta.versionId",
+            "Resource.text.div",
+            "Resource.text.status",
+        };
+
+        public JsonPatchService(IModelInfoProvider modelInfoProvider)
+        {
+            EnsureArg.IsNotNull(modelInfoProvider, nameof(modelInfoProvider));
+
+            _modelInfoProvider = modelInfoProvider;
+        }
+
+        public ResourceElement Patch(ResourceWrapper resourceToPatch, JsonPatchDocument patchDocument, WeakETag weakETag)
+        {
+            EnsureArg.IsNotNull(resourceToPatch, nameof(resourceToPatch));
+
+            Validate(resourceToPatch, weakETag);
+
+            var node = (FhirJsonNode)FhirJsonNode.Parse(resourceToPatch.RawResource.Data);
+
+            // Capture the state of properties that are immutable
+            ITypedElement resource = node.ToTypedElement(_modelInfoProvider.StructureDefinitionSummaryProvider);
+            (string path, object result)[] preState = _immutableProperties.Select(x => (path: x, result: resource.Scalar(x))).ToArray();
+
+            Resource patchedResource = GetPatchedJsonResource(node, patchDocument);
+
+            (string path, object result)[] postState = _immutableProperties.Select(x => (path: x, result: resource.Scalar(x))).ToArray();
+            if (!preState.Zip(postState).All(x => x.First.path == x.Second.path && string.Equals(x.First.result?.ToString(), x.Second.result?.ToString(), StringComparison.Ordinal)))
+            {
+                throw new RequestNotValidException(Core.Resources.PatchImmutablePropertiesIsNotValid);
+            }
+
+            return patchedResource.ToResourceElement();
+        }
+
+        private static void Validate(ResourceWrapper currentDoc, WeakETag eTag)
+        {
+            if (currentDoc.IsHistory)
+            {
+                throw new MethodNotAllowedException(Core.Resources.PatchVersionNotAllowed);
+            }
+
+            if (eTag != null && eTag.VersionId != currentDoc.Version)
+            {
+                throw new PreconditionFailedException(string.Format(Core.Resources.ResourceVersionConflict, eTag.VersionId));
+            }
+        }
+
+        private Resource GetPatchedJsonResource(FhirJsonNode node, JsonPatchDocument operations)
+        {
+            try
+            {
+                // Use low-level JSON parser
```

Comment:
Comment not correct here?

Reply:
get carried over from previous implementation. not sure what it means so removing.